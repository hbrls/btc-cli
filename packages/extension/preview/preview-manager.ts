import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as typescript from 'typescript';

import { SecurityPolicy } from '../security/security';

import {
  createOrShowPanel,
  refreshPanel,
} from './webview-manager';
import evaluateInWebview from './evaluate-in-webview';

const { performance, PerformanceObserver } = require('perf_hooks');

export interface StyleConfiguration {
  useVscodeMarkdownStyles: boolean;
  useWhiteBackground: boolean;
}

export interface TypeScriptConfiguration {
  tsCompilerOptions: typescript.CompilerOptions;
  tsCompilerHost: typescript.CompilerHost;
}

export let currentPreview: Preview | undefined;
export class Preview {
  /**
   * Current text document being previewed
   */
  doc: vscode.TextDocument;

  active: boolean;

  /**
   * Dependent doc being edited.
   * This is used to get doc text instead of reading from the file system,
   * when preview on change is configured.
   */
  editingDoc: vscode.TextDocument;

  /**
   * Dependent file paths
   */
  dependentFsPaths: Set<string>;

  /**
   * TODO: types
   * Curent Webview handle pertaining to this preview
   */
  webviewHandle: any;

  /**
   * Wait for webview handshake in order to use webview handle
   */
  webviewHandshakePromise: Promise<undefined>;
  resolveWebviewHandshakePromise: () => void;
  initWebviewHandshakePromise() {
    this.webviewHandshakePromise = new Promise(resolve => {
      this.resolveWebviewHandshakePromise = () => {
        resolve();
      };
    });
  }

  /**
   * User configuration: should update preview on change or update preview on save
   */
  configuration: {
    previewOnChange: boolean;
    useVscodeMarkdownStyles: boolean;
    useWhiteBackground: boolean;
    customLayoutFilePath: string;
    useSucraseTranspiler: boolean;
    securityPolicy: SecurityPolicy
  };

  typescriptConfiguration?: TypeScriptConfiguration;

  performanceObserver: PerformanceObserver;
  evaluationDuration: DOMHighResTimeStamp;
  previewDuration: DOMHighResTimeStamp;

  get styleConfiguration() {
    return {
      useVscodeMarkdownStyles: this.configuration.useVscodeMarkdownStyles,
      useWhiteBackground: this.configuration.useWhiteBackground
    };
  }

  get securityConfiguration() {
    return { securityPolicy: this.configuration.securityPolicy };
  }

  generateTypescriptConfiguration(configFile) {
    let tsCompilerOptions: typescript.CompilerOptions,
      tsCompilerHost: typescript.CompilerHost;
    if (configFile) {
      const configContents = fs.readFileSync(configFile).toString();
      const configJson = typescript.parseConfigFileTextToJson(
        configFile,
        configContents
      ).config.compilerOptions;
      tsCompilerOptions = typescript.convertCompilerOptionsFromJson(
        configJson,
        configFile
      ).options;
    } else {
      tsCompilerOptions = typescript.getDefaultCompilerOptions();
    }
    delete tsCompilerOptions.emitDeclarationOnly;
    delete tsCompilerOptions.declaration;
    tsCompilerOptions.module = typescript.ModuleKind.ESNext;
    tsCompilerOptions.target = typescript.ScriptTarget.ESNext;
    tsCompilerOptions.noEmitHelpers = false;
    tsCompilerOptions.importHelpers = false;
    tsCompilerHost = typescript.createCompilerHost(tsCompilerOptions);
    this.typescriptConfiguration = {
      tsCompilerHost,
      tsCompilerOptions
    };
  }

  constructor(doc: vscode.TextDocument) {
    this.setDoc(doc);
    const extensionConfig = vscode.workspace.getConfiguration('mdx-preview', doc.uri);
    this.configuration = {
      previewOnChange: extensionConfig.get<boolean>('preview.previewOnChange', true),
      useSucraseTranspiler: extensionConfig.get<boolean>('build.useSucraseTranspiler', false),
      useVscodeMarkdownStyles: extensionConfig.get<boolean>('preview.useVscodeMarkdownStyles', true),
      useWhiteBackground: extensionConfig.get<boolean>('preview.useWhiteBackground', false),
      customLayoutFilePath: extensionConfig.get<string>('preview.mdx.customLayoutFilePath', ""),
      securityPolicy: extensionConfig.get<SecurityPolicy>('preview.security', SecurityPolicy.Strict)
    };
    if (process.env.NODE_ENV === 'development') {
      this.performanceObserver = new PerformanceObserver(
        (list: PerformanceObserverEntryList, observer: PerformanceObserver) => {
          this.previewDuration = list.getEntries()[0].duration;
            vscode.window.showInformationMessage(`Previewing used: ${Number(this.previewDuration / 1000).toFixed(2)} seconds. 
            Evaluation used: ${ Number(this.evaluationDuration / 1000).toFixed(2) } seconds.`);
          performance.clearMarks();
        }
      );
      this.performanceObserver.observe({ entryTypes: ['measure'] });
    }
  }

  setDoc(doc: vscode.TextDocument) {
    this.doc = doc;

    this.dependentFsPaths = new Set([doc.uri.fsPath]);
    let configFile = typescript.findConfigFile(this.entryFsDirectory, typescript.sys.fileExists);
    if (configFile) {
      this.generateTypescriptConfiguration(configFile);
    } else {
      this.typescriptConfiguration = undefined;
    }
  }

  get fsPath():string {
    return this.doc.uri.fsPath;
  }

  get text():string {
    return this.doc.getText();
  }

  /**
   * Entry fs directory for resolving bare imports.
   * For untitled documents, we are trying to get the workspace root. If there's no workspace root,
   * we will not resolve any bare imports.
   * For file documents, it's directory that document is located.
   */
  get entryFsDirectory(): string | null {
    if (this.doc.uri.scheme === 'untitled') {
      const rootWorkspaceFolder = vscode.workspace.workspaceFolders[0];
      if (!rootWorkspaceFolder) {
        return null;
      }
      return rootWorkspaceFolder.uri.fsPath;
    } else if (this.doc.uri.scheme === 'file') {
      return path.dirname(this.fsPath);
    }
    return null;
  }

  updateWebview() {
    const preview = this;
    const { uri } = preview.doc;
    const { scheme, fsPath } = uri;
    switch (scheme) {
      case 'untitled': {
        evaluateInWebview(preview, preview.text, preview.entryFsDirectory);
        return;
      }
      case 'file': {
        if (this.configuration.previewOnChange) {
          evaluateInWebview(preview, preview.text, fsPath);
        } else {
          const text = fs.readFileSync(fsPath, { encoding: 'utf8' });
          evaluateInWebview(preview, text, fsPath);
        }
        break;
      }
      default:
        break;
    }
  }

  refreshWebview() {
    refreshPanel(currentPreview);
    this.updateWebview();
  }

  handleDidChangeTextDocument(fsPath: string, doc: vscode.TextDocument) {
    if (this.active) {
      if (this.configuration.previewOnChange) {
        if (this.dependentFsPaths.has(fsPath)) {
          this.editingDoc = doc;
          if (fsPath !== this.fsPath) {
            this.webviewHandle.invalidate(fsPath)
              .then(() => {
                this.updateWebview();
              });
          } else {
            // not necessary to invalidate entry path
            this.updateWebview();
          }
        }
      }
    }
  }

  handleDidSaveTextDocument(fsPath: string) {
    if (this.active) {
      if (this.dependentFsPaths.has(fsPath)) {
        if (fsPath !== this.fsPath) {
          this.webviewHandle.invalidate(fsPath)
            .then(() => {
              this.updateWebview();
            });
        } else {
          this.updateWebview();
        }
      }
    }
  }
  
  updateConfiguration() {
    const extensionConfig = vscode.workspace.getConfiguration('mdx-preview', this.doc.uri);
    const previewOnChange = extensionConfig.get<boolean>('preview.previewOnChange', true);
    const useSucraseTranspiler = extensionConfig.get<boolean>('build.useSucraseTranspiler', false);
    const useVscodeMarkdownStyles = extensionConfig.get<boolean>('preview.useVscodeMarkdownStyles', true);
    const useWhiteBackground = extensionConfig.get<boolean>('preview.useWhiteBackground', false);
    const customLayoutFilePath = extensionConfig.get<string>('preview.mdx.customLayoutFilePath', "");
    const securityPolicy = extensionConfig.get<SecurityPolicy>('preview.security', SecurityPolicy.Strict);

    const needsWebviewRefresh = useVscodeMarkdownStyles !== this.configuration.useVscodeMarkdownStyles
      || useWhiteBackground !== this.configuration.useWhiteBackground
      || customLayoutFilePath !== this.configuration.customLayoutFilePath
      || securityPolicy !== this.configuration.securityPolicy;

    Object.assign(this.configuration, {
      previewOnChange,
      useSucraseTranspiler,
      useVscodeMarkdownStyles,
      useWhiteBackground,
      customLayoutFilePath,
      securityPolicy
    });
    
    if (needsWebviewRefresh) {
      this.refreshWebview();
    }
  }
}

export function openPreview() {
  if (!vscode.window.activeTextEditor) {
    return;
  }
  const doc = vscode.window.activeTextEditor.document;
  if (!currentPreview) {
    currentPreview = new Preview(doc);
  } else {
    currentPreview.setDoc(doc);
  }
  createOrShowPanel(currentPreview);
  currentPreview.updateWebview();
}

export function refreshPreview() {
  if (!currentPreview) {
    return;
  }

  // don't set doc
  refreshPanel(currentPreview);
  currentPreview.updateWebview();
}