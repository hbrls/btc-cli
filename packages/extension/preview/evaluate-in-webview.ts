import * as fs from 'fs';

import { transformEntry } from '../module-fetcher/transform';

const precinct = require("precinct");

const { performance } = require('perf_hooks');
import { Preview } from './preview-manager';

/**
 * @param text text to preview
 * @param fsPath fsPath of current document
 */
export default async function evaluateInWebview(preview: Preview, text: string, fsPath: string) {
  const { webviewHandle } = preview;
  try {
    performance.mark('preview/start');

    const code = await transformEntry(text, fsPath, preview);
    const entryFilePath = fs.realpathSync(fsPath);
    const entryFileDependencies = precinct(code);

    console.log(code);
    console.log(entryFilePath);
    console.log(entryFileDependencies);

    await preview.webviewHandshakePromise;
    if (webviewHandle && webviewHandle.updatePreview) {
      webviewHandle.updatePreview(
        code,
        entryFilePath,
        entryFileDependencies
      );
    }
  } catch (error) {
    console.error(error);
    if (webviewHandle) {
      webviewHandle.showPreviewError({ 
        message: error.message
      });
    }
  }
}