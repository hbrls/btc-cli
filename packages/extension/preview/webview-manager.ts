import * as path from 'path';
import * as vscode from 'vscode';

import { Preview, StyleConfiguration } from './preview-manager';
import { getCSP } from '../security/CSP';
import { initRPCExtensionSide } from '../rpc-extension';

const VIEW_TYPE = 'mdx.preview';
const MDX_PREVIEW_FOCUS_CONTEXT_KEY = 'mdxPreviewFocus';

let panel: (vscode.WebviewPanel | undefined);
let panelDoc: (vscode.TextDocument | undefined);
let disposables: vscode.Disposable[] = [];

let webviewAppHTMLResources: {
  styleHref: string;
  runTimeMainSrc: string;
  chunkSrc: string;
  scriptSrc: string;
} | undefined;

export function initWebviewAppHTMLResources(context: vscode.ExtensionContext) {
    const { extensionPath } = context;
    const manifest = require(path.join(extensionPath, 'build', 'webview-app', 'asset-manifest.json'));
    const mainScript = manifest['main.js'];
    const mainStyle = manifest['main.css'];
    const scriptUri = vscode.Uri.file(path.join(extensionPath, 'build', 'webview-app', mainScript))
        .with({ scheme: 'vscode-resource' });
    const chunkPath = path.join(extensionPath, 'build', 'webview-app', manifest[Object.keys(manifest).find(key => key.endsWith('chunk.js'))]);
    const chunkUri = vscode.Uri.file(chunkPath).with({ scheme: 'vscode-resource' });
    const runTimeMainUri = vscode.Uri.file(path.join(extensionPath, 'build', 'webview-app', manifest['runtime~main.js']))
        .with({ scheme: 'vscode-resource' });
    const styleUri = vscode.Uri.file(path.join(extensionPath, 'build', 'webview-app', mainStyle))
        .with({ scheme: 'vscode-resource' });
    webviewAppHTMLResources = {
      styleHref: styleUri.toString(true),
      runTimeMainSrc: runTimeMainUri.toString(true),
      chunkSrc: chunkUri.toString(true),
      scriptSrc: scriptUri.toString(true)
    };
}

function getWebviewAppHTML(
    baseHref: string,
    contentSecurityPolicy: string,
    styleConfiguration: StyleConfiguration
) {
    if (!webviewAppHTMLResources) {
        return;
    }
    const {
      useVscodeMarkdownStyles,
      useWhiteBackground,
    } = styleConfiguration;
    let styleNodeHTML: string;
    let overrideBodyStyles = useWhiteBackground ? `body {
    color: black;
    background: white;
}`: '';
    // We have some styles attached in webview by vscode
    // If we don't use vscode's markdown styles, remove them
    let overrideDefaultStyles = (!useVscodeMarkdownStyles)
        ? `code {
  color: inherit;
}
blockquote {
  background: inherit;
}`: '';
    if (overrideBodyStyles || overrideDefaultStyles) {
        styleNodeHTML = `<style type="text/css">
  ${overrideBodyStyles}
  ${overrideDefaultStyles}
</style>`;
    } else {
        styleNodeHTML = '';
    }

    const { styleHref, runTimeMainSrc, chunkSrc, scriptSrc } = webviewAppHTMLResources;
    const webviewAppHTML = `<!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>MDX Preview</title>
        <link rel="stylesheet" type="text/css" href="${styleHref}">
        <meta http-equiv="Content-Security-Policy" content="${contentSecurityPolicy}">
        <base href="${baseHref}">
        ${styleNodeHTML}
    </head>
    <body>
        <div id="root"></div>
        <script crossorigin="anonymous" src="${runTimeMainSrc}"></script>
        <script crossorigin="anonymous" src="${chunkSrc}"></script>
        <script crossorigin="anonymous" src="${scriptSrc}"></script>
    </body>
</html>`;
    return webviewAppHTML;
}

function dispose() {
    panel.dispose();
    while (disposables.length) {
        const disposable = disposables.pop();
        if (disposable) {
            disposable.dispose();
        }
    }
    panel = undefined;
}

function setPanelHTMLFromPreview(preview: Preview) {
    const { doc, styleConfiguration, securityConfiguration } = preview;
    const previewBaseHref = doc.uri
        .with({ scheme: 'vscode-resource' })
        .toString(true);
    const webviewAppHTML = getWebviewAppHTML(previewBaseHref, getCSP(securityConfiguration), styleConfiguration);
    panel.webview.html = webviewAppHTML;
}

export function createOrShowPanel(preview: Preview) {
    const activeTextEditorColumn = vscode.window.activeTextEditor ? vscode.window.activeTextEditor.viewColumn : vscode.ViewColumn.One;
    const previewColumn = activeTextEditorColumn + 1;// vscode.ViewColumn.Beside || vscode.ViewColumn.One;
    const previewTitle = `Preview ${path.basename(preview.doc.fileName)}`;

    if (!panel) {
        panel = vscode.window.createWebviewPanel(
            VIEW_TYPE,
            previewTitle,
            previewColumn,
            {
                enableScripts: true,
                enableCommandUris: true,
                retainContextWhenHidden: true
            }
        );
        panelDoc = preview.doc;
        setPanelHTMLFromPreview(preview);

        vscode.commands.executeCommand('setContext', MDX_PREVIEW_FOCUS_CONTEXT_KEY, true);

        panel.onDidDispose(() => {
            preview.active = false;
            dispose();
        }, null, disposables);
        panel.onDidChangeViewState(({ webviewPanel }) => {
            vscode.commands.executeCommand('setContext', MDX_PREVIEW_FOCUS_CONTEXT_KEY, webviewPanel.active);
        }, null, disposables);
        
        preview.initWebviewHandshakePromise();
        const webviewHandle = initRPCExtensionSide(preview, panel.webview, disposables);
        preview.webviewHandle = webviewHandle;
    } else {
        if (panelDoc !== preview.doc) {
            panel.title = previewTitle;
            setPanelHTMLFromPreview(preview);
            panelDoc = preview.doc;
        }
        panel.reveal(previewColumn);

        vscode.commands.executeCommand('setContext', MDX_PREVIEW_FOCUS_CONTEXT_KEY, true);
    }
    preview.active = true;
    return panel;
}

export function refreshPanel(preview: Preview) {
    if (!panel) {
        return;
    } else {
        // reveal in current column, and preserves focus
        panel.reveal(undefined, true);
        panel.webview.html = '';
        setPanelHTMLFromPreview(preview);
    }
}