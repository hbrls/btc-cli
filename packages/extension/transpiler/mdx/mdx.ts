import mdx from '@mdx-js/mdx';
import hasDefaultExport from './hasDefaultExport';
import * as path from 'path';

import { Preview } from '../../preview/preview-manager';

const injectMDXStyles = (mdxText: string, preview: Preview):string => {
  const {
    customLayoutFilePath,
    useVscodeMarkdownStyles,
    useWhiteBackground,
  } = preview.configuration;

  if (customLayoutFilePath) {
    try {
      const currentPreviewDirname = path.dirname(preview.doc.uri.fsPath);
      const relativeCustomLayoutPath = path.relative(currentPreviewDirname, customLayoutFilePath);
      return `import Layout from '.${path.sep}${relativeCustomLayoutPath}';

export default Layout;

${mdxText}`;
    } catch (error) {
      return mdxText;
    }
  } else if (useVscodeMarkdownStyles) {
    let layoutOptions = useWhiteBackground ? '{ forceLightTheme: true }' : '{}';
    return `import { createLayout } from 'vscode-markdown-layout';

export default createLayout(${layoutOptions});

${mdxText}`;
  } else {
    return mdxText;
  }
};

const wrapCompiledMdx = (compiledMDX: string, isEntry: boolean) => {
  if (isEntry) {
    // entry MDX, render on webview DOM
    return `import React from 'react';
import ReactDOM from 'react-dom';
import { MDXTag } from '@mdx-js/tag';
${compiledMDX}
ReactDOM.render(<MDXContent></MDXContent>, document.getElementById('root'));`;
  } else {
    // transclusion
    return `import React from 'react';
import { MDXTag } from '@mdx-js/tag';
${compiledMDX}`;
  }
};

export const mdxTranspileAsync = async (mdxText: string, isEntry: boolean, preview: Preview) => {
  let mdxTextToCompile: string;
  if (!hasDefaultExport(mdxText)) {
    // inject vscode markdown styles if we haven't found a default export
    mdxTextToCompile = injectMDXStyles(mdxText, preview);
  } else {
    mdxTextToCompile = mdxText;
  }
  const compiledMDX = await mdx(mdxTextToCompile);
  return wrapCompiledMdx(compiledMDX, isEntry);
};
