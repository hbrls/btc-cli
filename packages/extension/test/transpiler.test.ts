import * as vscode from 'vscode';
import { InMemoryDocument } from './InMemoryDocument';
import { Preview } from '../preview/preview-manager';

import * as mdx from '../transpiler/mdx/mdx';

describe("Transpiler Tests", function () {

  test("Transpiles entry mdx file (no default export)", async function () {
    const content = 'hello';
    const mockDoc = new InMemoryDocument(vscode.Uri.file('test.md'), content);
    const mockPreview = new Preview(mockDoc);
    const mdxText = mockPreview.text;
    const isEntry = true;
    const compiledMdx = await mdx.mdxTranspileAsync(mdxText, isEntry, mockPreview);
    expect(compiledMdx).toMatchSnapshot();
  });

  // with white background config adds the necessary params to layout
  test("Transpiles entry mdx file with vscode markdown styles and white background config (no default export)", async function () {
    const content = 'hello';
    const mockDoc = new InMemoryDocument(vscode.Uri.file('test.md'), content);
    const mockPreview = new Preview(mockDoc);
    mockPreview.configuration.useVscodeMarkdownStyles = true;
    mockPreview.configuration.useWhiteBackground = true;
    const mdxText = mockPreview.text;
    const isEntry = true;
    const compiledMdx = await mdx.mdxTranspileAsync(mdxText, isEntry, mockPreview);
    expect(compiledMdx).toMatchSnapshot();
  });

  test("Transpiles entry mdx file with no vscode markdown styles (no default export)", async function () {
    const content = 'hello';
    const mockDoc = new InMemoryDocument(vscode.Uri.file('test.md'), content);
    const mockPreview = new Preview(mockDoc);
    mockPreview.configuration.useVscodeMarkdownStyles = false;
    const mdxText = mockPreview.text;
    const isEntry = true;
    const compiledMdx = await mdx.mdxTranspileAsync(mdxText, isEntry, mockPreview);
    expect(compiledMdx).toMatchSnapshot();
  });

  // custom layout file takes precedence over vscode markdown styles
  test("Transpiles entry mdx file with custom layout file (no default export)", async function () {
    const content = 'hello';
    const mockDoc = new InMemoryDocument(vscode.Uri.file('test.md'), content);
    const mockPreview = new Preview(mockDoc);
    mockPreview.configuration.customLayoutFilePath = '/mdx/customLayout.js';
    mockPreview.configuration.useVscodeMarkdownStyles = true;
    const mdxText = mockPreview.text;
    const isEntry = true;
    const compiledMdx = await mdx.mdxTranspileAsync(mdxText, isEntry, mockPreview);
    expect(compiledMdx).toMatchSnapshot();
  });

  // non entry file don't append render code to the DOM
  test("Transpiles non-entry mdx file (no default export)", async function () {
    const content = 'hello';
    const mockDoc = new InMemoryDocument(vscode.Uri.file('test.md'), content);
    const mockPreview = new Preview(mockDoc);
    const mdxText = mockPreview.text;
    const isEntry = false;
    const compiledMdx = await mdx.mdxTranspileAsync(mdxText, isEntry, mockPreview);
    expect(compiledMdx).toMatchSnapshot();
  });

  // default export takes precedence over custom layout file or vscode markdown styles
  test("Transpiles entry mdx file (with default export)", async function () {
    const content = `import Layout from './layout';\n\nexport default Layout;\n\nhello`;
    const mockDoc = new InMemoryDocument(vscode.Uri.file('test.md'), content);
    const mockPreview = new Preview(mockDoc);
    mockPreview.configuration.customLayoutFilePath = '/mdx/customLayout.js';
    mockPreview.configuration.useVscodeMarkdownStyles = true;
    const mdxText = mockPreview.text;
    const isEntry = true;
    const compiledMdx = await mdx.mdxTranspileAsync(mdxText, isEntry, mockPreview);
    expect(compiledMdx).toMatchSnapshot();
  });

});
