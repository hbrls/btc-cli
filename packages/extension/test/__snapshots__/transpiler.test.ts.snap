// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transpiler Tests Transpiles entry mdx file (no default export) 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MDXTag } from '@mdx-js/tag';
import { createLayout } from 'vscode-markdown-layout';

const layoutProps = {
  
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props)
    this.layout = createLayout({})
  }
  render() {
    const { components, ...props } = this.props

    return <MDXTag
             name=\\"wrapper\\"
             Layout={this.layout} layoutProps={Object.assign({}, layoutProps, props)}
             components={components}>

<MDXTag name=\\"p\\" components={components}>{\`hello\`}</MDXTag>
           </MDXTag>
  }
}
MDXContent.isMDXComponent = true
ReactDOM.render(<MDXContent></MDXContent>, document.getElementById('root'));"
`;

exports[`Transpiler Tests Transpiles entry mdx file (with default export) 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MDXTag } from '@mdx-js/tag';
import Layout from './layout';

const layoutProps = {
  
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props)
    this.layout = Layout
  }
  render() {
    const { components, ...props } = this.props

    return <MDXTag
             name=\\"wrapper\\"
             Layout={this.layout} layoutProps={Object.assign({}, layoutProps, props)}
             components={components}>

<MDXTag name=\\"p\\" components={components}>{\`hello\`}</MDXTag>
           </MDXTag>
  }
}
MDXContent.isMDXComponent = true
ReactDOM.render(<MDXContent></MDXContent>, document.getElementById('root'));"
`;

exports[`Transpiler Tests Transpiles entry mdx file with custom layout file (no default export) 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MDXTag } from '@mdx-js/tag';
import Layout from './mdx/customLayout.js';

const layoutProps = {
  
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props)
    this.layout = Layout
  }
  render() {
    const { components, ...props } = this.props

    return <MDXTag
             name=\\"wrapper\\"
             Layout={this.layout} layoutProps={Object.assign({}, layoutProps, props)}
             components={components}>

<MDXTag name=\\"p\\" components={components}>{\`hello\`}</MDXTag>
           </MDXTag>
  }
}
MDXContent.isMDXComponent = true
ReactDOM.render(<MDXContent></MDXContent>, document.getElementById('root'));"
`;

exports[`Transpiler Tests Transpiles entry mdx file with no vscode markdown styles (no default export) 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MDXTag } from '@mdx-js/tag';


const layoutProps = {
  
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props)
    this.layout = null
  }
  render() {
    const { components, ...props } = this.props

    return <MDXTag
             name=\\"wrapper\\"
             
             components={components}><MDXTag name=\\"p\\" components={components}>{\`hello\`}</MDXTag>
           </MDXTag>
  }
}
MDXContent.isMDXComponent = true
ReactDOM.render(<MDXContent></MDXContent>, document.getElementById('root'));"
`;

exports[`Transpiler Tests Transpiles entry mdx file with vscode markdown styles and white background config (no default export) 1`] = `
"import React from 'react';
import ReactDOM from 'react-dom';
import { MDXTag } from '@mdx-js/tag';
import { createLayout } from 'vscode-markdown-layout';

const layoutProps = {
  
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props)
    this.layout = createLayout({ forceLightTheme: true })
  }
  render() {
    const { components, ...props } = this.props

    return <MDXTag
             name=\\"wrapper\\"
             Layout={this.layout} layoutProps={Object.assign({}, layoutProps, props)}
             components={components}>

<MDXTag name=\\"p\\" components={components}>{\`hello\`}</MDXTag>
           </MDXTag>
  }
}
MDXContent.isMDXComponent = true
ReactDOM.render(<MDXContent></MDXContent>, document.getElementById('root'));"
`;

exports[`Transpiler Tests Transpiles non-entry mdx file (no default export) 1`] = `
"import React from 'react';
import { MDXTag } from '@mdx-js/tag';
import { createLayout } from 'vscode-markdown-layout';

const layoutProps = {
  
};
export default class MDXContent extends React.Component {
  constructor(props) {
    super(props)
    this.layout = createLayout({})
  }
  render() {
    const { components, ...props } = this.props

    return <MDXTag
             name=\\"wrapper\\"
             Layout={this.layout} layoutProps={Object.assign({}, layoutProps, props)}
             components={components}>

<MDXTag name=\\"p\\" components={components}>{\`hello\`}</MDXTag>
           </MDXTag>
  }
}
MDXContent.isMDXComponent = true"
`;
